mpack-instance-manager/src/main/python/instance_manager/instance_manager.py

```diff
@@ -0,0 +1,500 @@
+regarding copyright ownership.  The ASF licenses this file
+to you under the Apache License, Version 2.0 (the
+"License"); you may not use this file except in compliance
+with the License.  You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+
+"""
+
+__all__ = ["create_mpack", "set_mpack_instance", "get_conf_dir", "list_instances"]
+
+import sys
+import os
+import json
+
+MPACK_JSON_FILE_NAME = 'mpack.json'
+CURRENT_SOFTLINK_NAME = 'current'
+CONFIGS_DIRECTORY_NAME = 'conf'
+
+ROOT_FOLDER_PATH = "/opt/odp"
+
+INSTANCES_FOLDER_NAME = "instances"
+MODULES_FOLDER_NAME = "modules"
+MPACKS_FOLDER_NAME = "mpacks"
+DEFAULT_COMPONENT_INSTANCE_NAME = 'default'
+DEFAULT_MPACK_INSTANCE_NAME = 'default'
+
+
+def create_mpack(mpack_name, mpack_version, mpack_instance, subgroup_name, module_name,
+                 components=None, components_map=None):
+  """
+  Use case 1: Creates an instance of mpack with a new subgroup, new module and one or more component(s)
+
+  Use case 2: Creates/adds in an existing mpack either a : new subgroup and/or module and/or component instance.
+
+  Components are provided as map with key as 'component type' and value as 'list of individual component instances
+  names' OR empty map to create single instance of all components with name 'default'
+  OR
+  list of 'components_instances_name' to be created. (default) OR '*' for all components
+  """
+  MpackInstance.create_mpack_instance(mpack_name, mpack_version, mpack_instance, subgroup_name, module_name,
+                                      components, components_map)
+
+
+def set_mpack_instance(mpack, mpack_version, mpack_instance, subgroup_name, module_name,
+                       components=None, components_map=None):
+  """
+  Use case: move a given component instances from one version to the next version by modifying the soft links.
+  Eg: Moving hive_server instance 'HS-1' from version 1.0.0-b1 to 1.5.0-b1.
+
+  Components are provided as map with key as 'component type' and value as 'list of individual component instances
+  names' OR empty map to create single instance of all components with name 'default'
+  OR
+  list of 'components_instances_name' to be created. (default) OR '*' for all components
+  """
+  instances = MpackInstance.parse_instances_with_filtering(os.path.join(ROOT_FOLDER_PATH, INSTANCES_FOLDER_NAME), mpack,
+                                                           mpack_instance, subgroup_name, module_name,
+                                                           components, components_map)
+  if not instances:
+    print("\nFound no instances for the given filters.")
+    sys.exit(0)
```

Comment:
We should only throw exceptions in the instance manager module and not sys.exit(). The wrapper script can gracefully exit.